webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala

```diff
@@ -275,12 +278,25 @@ trait CachedFindRuleNodeStatusReports extends ReportingService with CachedReposi
   val updateCacheFromRequest: IO[Nothing, Unit] = invalidateComplianceRequest.take.flatMap(invalidatedIds =>
     // batch node processing by slice of batchSize.
     // Be careful, sliding default step is 1.
-    ZIO.foreach_(invalidatedIds.sliding(batchSize,batchSize).to(Iterable))(nodeIds =>
-      (for {
-        updated <- defaultFindRuleNodeStatusReports.findRuleNodeStatusReports(nodeIds.toSet, Set()).toIO
-        _       <- IOResult.effectNonBlocking { cache = cache ++ updated }
-        _       <- ReportLoggerPure.Cache.debug(s"Compliance cache updated for nodes: ${nodeIds.map(_.value).mkString(", ")}")
-      } yield ()).catchAll(err => ReportLoggerPure.Cache.error(s"Error when updating compliance cache for nodes: [${nodeIds.map(_.value).mkString(", ")}]: ${err.fullMsg}"))
+
+    ZIO.foreach_(invalidatedIds.sliding(batchSize,batchSize).to(Iterable))(nodeIdsWithOptionnalCompliance =>
+      // list is limited in size by construction.to batchSize
+      // i need to take the last one for each nodeid (if compliance is invalidated by a generation after its been computed, it still
+      // needs to be invalidated
+      // so groupBy NodeId, and take last element for each
+
+      {
+        val (nodeIdsWithoutCompliance, nodeIdsWithCompliance) = nodeIdsWithOptionnalCompliance.groupBy(_._1).map { case (nodeId, list) => (nodeId, list.last._2) }.partition(_._2.isDefined)
+
+        for {
+          updated <- defaultFindRuleNodeStatusReports.findRuleNodeStatusReports(nodeIdsWithoutCompliance.keys.toSet, Set()).toIO
+          _       <- IOResult.effectNonBlocking {
+            cache = cache ++ updated
```

Comment:
cache ought to be more complete with
NodeId, LastRunDate, ExpectedConfiguration, NodeConfigurationStatus